pkg/common/file.go

```diff
@@ -34,177 +36,147 @@ type Config struct {
+	debugKey             = "debug"
+	pullImageOnCreateKey = "pull-image-on-create"
+	disablePullOnRunKey  = "disable-pull-on-run"
+)
+
 // ReadConfig reads from a file with the given name and returns a config or
 // an error if the file was unable to be parsed.
-func ReadConfig(filepath string) (*Config, error) {
-	data, err := os.ReadFile(filepath)
+func ReadConfig(path string) (*Config, error) {
+	data, err := os.ReadFile(path)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("read config file path: %w", err)
+	}
+
+	m := map[string]any{}
+	if err := yaml.Unmarshal(data, &m); err != nil {
+		return nil, fmt.Errorf("unmarshal YAML config: %w", err)
+	}
+	logrus.Debugf("Unmarshaled config map: %v", m)
+
+	for k := range m {
+		if !slices.Contains([]string{
+			runtimeEndpointKey,
+			imageEndpointKey,
+			timeoutKey,
+			debugKey,
+			pullImageOnCreateKey,
+			disablePullOnRunKey,
+		}, k) {
+			return nil, fmt.Errorf("invalid config option: %s", k)
+		}
 	}
-	yamlConfig := &yaml.Node{}
-	err = yaml.Unmarshal(data, yamlConfig)
+
+	c := &Config{}
+
+	runtimeEndpoint, err := mapKeyValue[string](m, runtimeEndpointKey)
 	if err != nil {
 		return nil, err
 	}
-	config, err := getConfigOptions(yamlConfig)
+	c.RuntimeEndpoint = runtimeEndpoint
+
+	imageEndpoint, err := mapKeyValue[string](m, imageEndpointKey)
 	if err != nil {
 		return nil, err
 	}
-	return config, err
-}
+	c.ImageEndpoint = imageEndpoint
 
-// WriteConfig writes config to file
-// an error if the file was unable to be written to.
-func WriteConfig(c *Config, filepath string) error {
-	if c == nil {
-		c = &Config{}
-	}
-	if c.yamlData == nil {
-		c.yamlData = &yaml.Node{}
+	timeout, err := mapKeyValue[int](m, timeoutKey)
+	if err != nil {
+		return nil, err
 	}
+	c.Timeout = timeout
 
-	setConfigOptions(c)
+	debug, err := mapKeyValue[bool](m, debugKey)
+	if err != nil {
+		return nil, err
+	}
+	c.Debug = debug
 
-	data, err := yaml.Marshal(c.yamlData)
+	pullImageOnCreate, err := mapKeyValue[bool](m, pullImageOnCreateKey)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	c.PullImageOnCreate = pullImageOnCreate
 
-	if err := os.MkdirAll(gofilepath.Dir(filepath), 0o755); err != nil {
-		return err
+	disablePullOnRun, err := mapKeyValue[bool](m, disablePullOnRunKey)
+	if err != nil {
+		return nil, err
 	}
-	return os.WriteFile(filepath, data, 0o644)
+	c.DisablePullOnRun = disablePullOnRun
+
+	return c, nil
 }
 
-// Extracts config options from the yaml data which is loaded from file.
-func getConfigOptions(yamlData *yaml.Node) (*Config, error) {
-	config := &Config{yamlData: yamlData}
-
-	if len(yamlData.Content) == 0 ||
-		yamlData.Content[0].Content == nil {
-		return config, nil
-	}
-	contentLen := len(yamlData.Content[0].Content)
-
-	// YAML representation contains 2 yaml ScalarNodes per config option.
-	// One is config option name and other is the value of the option
-	// These ScalarNodes help preserve comments associated with
-	// the YAML entry
```

Comment:
I don't understand this comment. Why and where the comments associated with the YAML entry was ever used?

Reply:
besides this comment it is lgtm

Reply:
Looks like we cannot preserve the comments without the current logic. Closing this PR for now.

Reply:
I am not sure where we are preserving comments. Aren't config is just readonly?

If we do preserver comments in some scenario - it will be great to have a test for it so it is not just a comment.

Reply:
We preserve the comments but not the newlines right now, I added an e2e test scenario for the whole `crictl config` subcommand in https://github.com/kubernetes-sigs/cri-tools/pull/1659